### PR TITLE
Pregnancy visit test

### PR DESCRIPTION
**Pull request by ngaruko**
# Description

e2e for pregnancy visit report.

medic/cht-core#[number]

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/e2e/forms/pregnancy-visit.wdio-spec.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+const utils = require('../../utils');
+const userData = require('../../page-objects/forms/data/user.po.data');
+const loginPage = require('../../page-objects/login/login.wdio.page');
+const commonPage = require('../../page-objects/common/common.wdio.page');
+const reportsPage = require('../../page-objects/reports/reports.wdio.page');
+const genericForm = require('../../page-objects/forms/generic-form.wdio.page');
+const pregnancyVisitForm = require('../../page-objects/forms/pregnancy-visit-form.wdio.page');
+const { expect } = require('chai');
+const xml = fs.readFileSync(`${__dirname}/../../../config/standard/forms/app/pregnancy_visit.xml`, 'utf8');
```
Comment (dianabarsan) [inline tests/e2e/forms/pregnancy-visit.wdio-spec.js:12]: I wonder if we should move the while xml + form document generation into the .page file.

File: tests/e2e/forms/pregnancy-visit.wdio-spec.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+const fs = require('fs');
+
+const utils = require('../../utils');
+const userData = require('../../page-objects/forms/data/user.po.data');
+const loginPage = require('../../page-objects/login/login.wdio.page');
+const commonPage = require('../../page-objects/common/common.wdio.page');
+const reportsPage = require('../../page-objects/reports/reports.wdio.page');
+const genericForm = require('../../page-objects/forms/generic-form.wdio.page');
+const pregnancyVisitForm = require('../../page-objects/forms/pregnancy-visit-form.wdio.page');
+const { expect } = require('chai');
```
Comment (dianabarsan) [inline tests/e2e/forms/pregnancy-visit.wdio-spec.js:11]: I think global expect is already available. Does it not work as expected without this import?

File: tests/e2e/forms/pregnancy-visit.wdio-spec.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  it('Submit and validate Pregnancy Visit form and keeps the report minified', async () => {
+    await reportsPage.openForm('Pregnancy Visit');
+    await pregnancyVisitForm.selectPatient('jack');
+    await genericForm.nextPage();
+    const selectedDangerSigns = await pregnancyVisitForm.selectAllDangerSigns();
+    await genericForm.nextPage();
+    pregnancyVisitForm.addNotes;
+    await genericForm.nextPage();
+
+    await expect(await pregnancyVisitForm.dangerSignLabel().getText()).to.equal('Danger Signs');
```
Comment (dianabarsan) [inline tests/e2e/forms/pregnancy-visit.wdio-spec.js:43]: I don't think you need to `await expect`.

File: tests/page-objects/forms/generic-form.wdio.page.js [hunk 4]
```diff
@@ -23,6 +26,7 @@ const invalidateReport = async () => {
 };
 
 const validateReport = async () => {
+  console.log('validating reports...');
```
Comment (dianabarsan) [inline tests/page-objects/forms/generic-form.wdio.page.js:29]: Please remove this debug log.

File: tests/page-objects/forms/pregnancy-visit-form.wdio.page.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+const genericForm = require('./generic-form.wdio.page');
+const dangerSignLabel  =  () =>
```
Comment (dianabarsan) [inline tests/page-objects/forms/pregnancy-visit-form.wdio.page.js:2]: I think there are some extra spaces here. Also applies to next couple of lines. 
```suggestion
const dangerSignLabel = () =>
```

File: tests/page-objects/forms/pregnancy-visit-form.wdio.page.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+const selectDangerSign = async (value) => {
+  const danger = await $(`input[value="${value}"]`);
+  await danger.click();
+};
+const selectAllDangerSigns = async () => {
+  const checkboxes = await $$('input[name="/pregnancy_visit/group_danger_signs/g_danger_signs"]');
+  for (const checkbox of checkboxes) {
+    await checkbox.click();
+  }
+  return await checkboxes.length;
```
Comment (dianabarsan) [inline tests/page-objects/forms/pregnancy-visit-form.wdio.page.js:21]: I think you don't need this `await`

File: tests/page-objects/reports/reports.wdio.page.js [hunk 7]
```diff
@@ -64,6 +65,7 @@ const openForm = async (name) => {
   for (const form of await forms()) {
     if (await form.getText() === name) {
       await form.click();
+      await browser.refresh();
```
Comment (dianabarsan) [inline tests/page-objects/reports/reports.wdio.page.js:68]: I don't think this should be necessary.

File: tests/e2e/forms/pregnancy-visit.wdio-spec.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    await commonPage.goToReports();
+  });
+
+  it('Submit and validate Pregnancy Visit form and keeps the report minified', async () => {
+    await reportsPage.openForm('Pregnancy Visit');
+    await pregnancyVisitForm.selectPatient('jack');
+    await genericForm.nextPage();
+    const selectedDangerSigns = await pregnancyVisitForm.selectAllDangerSigns();
+    await genericForm.nextPage();
+    pregnancyVisitForm.addNotes;
```
Comment (dianabarsan) [inline tests/e2e/forms/pregnancy-visit.wdio-spec.js:40]:
```suggestion
await pregnancyVisitForm.addNotes();
```

File: tests/e2e/forms/pregnancy-visit.wdio-spec.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    await genericForm.nextPage();
+
+    await expect(await pregnancyVisitForm.dangerSignLabel().getText()).to.equal('Danger Signs');
+    await expect(await pregnancyVisitForm.dangerSignSummary().length).to.equal(selectedDangerSigns + 1);
+    await expect(await pregnancyVisitForm.followUpMessage().getText())
+      .to.have.string('Please note that Jack has one or more danger signs for a high risk pregnancy');
+    await reportsPage.submitForm();
+
+    await genericForm.verifyReport();
```
Comment (dianabarsan) [inline tests/e2e/forms/pregnancy-visit.wdio-spec.js:49]: I'm curious why report validation is part of this e2e test. 
Wouldn't it be more suitable to actually check the contents of the submitted report?

Comment (ngaruko) [reply tests/e2e/forms/pregnancy-visit.wdio-spec.js]: added some report checks

**Review by dianabarsan — APPROVED** [review]
Awesome!

**Pull request merged**